numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -843,29 +971,24 @@ VOID_setitem(PyObject *op, void *input, void *vap)
         return res;
     }
 
-    /* Default is to use buffer interface to set item */
+    /*
+     * Fall through case - non-structured void datatype. This is a very
+     * undiscerning case: It interprets any object as a buffer
+     * and reads as many bytes as possible, padding with 0.
+     */
     {
         const void *buffer;
         Py_ssize_t buflen;
-        if (PyDataType_FLAGCHK(descr, NPY_ITEM_HASOBJECT)
-                || PyDataType_FLAGCHK(descr, NPY_ITEM_IS_POINTER)) {
-            PyErr_SetString(PyExc_ValueError,
-                    "Setting void-array with object members using buffer.");
-            return -1;
-        }
```

Comment:
Because all cases involving objects now must fall into the HASOBJECTS branch. Because of other work I've done I'm fairly sure it is not possible to create a void unstructured type containing pointers.

In the old code, however, many structured arrays would fall into this part of the code (eg, see #6314).